core/src/test/scala/unit/kafka/log/LogLoaderTest.scala

```diff
@@ -352,7 +353,10 @@ class LogLoaderTest {
       // Intercept all segment read calls
       val interceptedLogSegments = new LogSegments(topicPartition) {
         override def add(segment: LogSegment): LogSegment = {
-          val wrapper = new LogSegment(segment) {
+          val idx = LazyIndex.forOffset(LogFileUtils.offsetIndexFile(wrapperLogDir, segment.baseOffset()), segment.baseOffset(), 1000)
```

Comment:
How about using `Mockito.spy`? for example:

```scala
          val wrapper = Mockito.spy(segment)
          Mockito.doAnswer{ in =>
            segmentsWithReads += wrapper
            segment.read(in.getArgument(0, classOf[java.lang.Long]), in.getArgument(1, classOf[java.lang.Integer]), in.getArgument(2, classOf[java.lang.Long]), in.getArgument(3, classOf[java.lang.Boolean]))
          }.when(wrapper).read(ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())
          Mockito.doAnswer{ in =>
            recoveredSegments += wrapper
            segment.recover(in.getArgument(0, classOf[ProducerStateManager]), in.getArgument(1, classOf[Optional[LeaderEpochFileCache]]))
          }.when(wrapper).recover(ArgumentMatchers.any(), ArgumentMatchers.any())
          super.add(wrapper)
```

Reply:
thanks for the suggestion! let me check that

Reply:
learnt a new framework for testing, thanks for sharing this!